pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/statement/JsonStatementOptimizer.java

```diff
@@ -0,0 +1,561 @@
+        optimizeJsonIdentifier(expression, schema, DataSchema.ColumnDataType.STRING);
+      }
+    }
+
+    // In HAVING clause, replace JSON path expressions with JSON_EXTRACT_SCALAR. This expression must match the
+    // corresponding SELECT list expression except for the alias.
+    Expression expression = query.getHavingExpression();
+    if (expression != null) {
+      optimizeJsonIdentifier(expression, schema, DataSchema.ColumnDataType.STRING);
+    }
+  }
+
+  /**
+   * Replace an json path expression with an aliased JSON_EXTRACT_SCALAR function.
+   * @param expression input expression to rewrite into JSON_EXTRACT_SCALAR function if the expression is json path.
+   * @param outputDataType to keep track of output datatype of JSON_EXTRACT_SCALAR function which depends upon the outer
+   *                 function that json path expression appears in.
+   * @return A {@link Pair} of values where the first value is alias for the input expression and second
+   * value indicates whether json path expression was found (true) or not (false) in the expression.
+   */
+  private static Pair<String, Boolean> optimizeJsonIdentifier(Expression expression, @Nullable Schema schema,
+      DataSchema.ColumnDataType outputDataType) {
+    switch (expression.getType()) {
+      case LITERAL:
+        return new Pair<>(getLiteralSQL(expression.getLiteral(), true), false);
+      case IDENTIFIER: {
+        String[] parts = getIdentifierParts(expression.getIdentifier());
+        boolean hasJsonPathExpression = false;
+        String alias = expression.getIdentifier().getName();
+        if (parts.length > 1 && isValidJSONColumn(parts[0], schema)) {
+          // replace <column-name>.<json-path> with json_extract_scalar(<column-name>, '<json-path>', 'STRING', <JSON-null-value>)
+          Function jsonExtractScalarFunction = getJsonExtractFunction(parts, outputDataType);
+          expression.setIdentifier(null);
+          expression.setType(ExpressionType.FUNCTION);
+          expression.setFunctionCall(jsonExtractScalarFunction);
+          hasJsonPathExpression = true;
+        }
+        return new Pair<>(alias, hasJsonPathExpression);
+      }
+      case FUNCTION: {
+        Function function = expression.getFunctionCall();
+        List<Expression> operands = function.getOperands();
+
+        boolean hasJsonPathExpression = false;
+        StringBuffer alias = new StringBuffer();
+        if (function.getOperator().toUpperCase().equals("AS")) {
+          // We don't need to compute an alias for AS function since AS function defines its own alias.
+          hasJsonPathExpression = optimizeJsonIdentifier(operands.get(0), schema, outputDataType).getSecond();
+          alias.append(function.getOperands().get(1).getIdentifier().getName());
+        } else {
+          // For all functions besides AS function, process the operands and compute the alias.
+          alias.append(function.getOperator().toLowerCase(Locale.ROOT)).append("(");
```

Comment:
Why `Locale.ROOT`?  I believe we use local (`toLowerCase()`) in other places, so better to keep it consistent

Reply:
Fixed.